### PR TITLE
added ui option to edit ni device name

### PR DESCRIPTION
**Pull request by NikitaPoly**

### CODE REVIEW

File: src/controllers/NI/src/NIDAQmx.cpp [hunk 1]
```diff
@@ -301,6 +301,35 @@ bool NIDAQmx::WriteDigitalLines(
...
+
+    if(DAQmxFailed(DAQmxGetSysDevNames(devicenamesbuffer, bufferSize))){
+        spdlog::error("NI device search failed");
+        std::vector<std::string> empty;
+        return empty;
+    }else{
+        std::string tempstring(devicenamesbuffer);
+        std::vector<std::string> devicenamesbuffer;
+        std::stringstream ss("Dev1,Dev2,Dev3");
+        while(ss.good()){
```
Comment (jperson) [inline src/controllers/NI/src/NIDAQmx.cpp:319]: can you explain what is going on here?

File: src/controllers/NI/src/NIDAQmx.cpp [hunk 2]
```diff
@@ -301,6 +301,35 @@ bool NIDAQmx::WriteDigitalLines(
...
+    char devicenamesbuffer[bufferSize] = {};
+
+    if(DAQmxFailed(DAQmxGetSysDevNames(devicenamesbuffer, bufferSize))){
+        spdlog::error("NI device search failed");
+        std::vector<std::string> empty;
+        return empty;
+    }else{
+        std::string tempstring(devicenamesbuffer);
+        std::vector<std::string> devicenamesbuffer;
+        std::stringstream ss("Dev1,Dev2,Dev3");
```
Comment (jperson) [inline src/controllers/NI/src/NIDAQmx.cpp:318]: are there only ever 3 devices, I don't understand what this is for

Comment (NikitaPoly) [reply src/controllers/NI/src/NIDAQmx.cpp]: This is left over from testing. I don't have any devices so I hardcoded what the output would look like to test my code. I updated the code now.

File: src/app/src/advancedsetupdialog.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+* Instance of advanced setup options window.
+*
+* @param parent Pointer to parent widget.
+*/
+AdvancedSetupDialog::AdvancedSetupDialog( toml::value config,std::string* m_niDev,QWidget *parent) : QDialog(parent), ui(new Ui::AdvancedSetupDialog) {
+    ui->setupUi(this);
+
+    std::string niDev = toml::find_or<std::string>(config, "device", "nidaqmx","device", std::string("Dev2"));
+    new_niDev = niDev;
+    m_niDev=m_niDev;
```
Comment (jperson) [inline src/app/src/advancedsetupdialog.cpp:17]: if you're passing in the config and reading the device name I don't think you need to also pass in the m_niDev, this class also shouldn't modify the `m_niDev` that is passed in, it should let `MainWindow` do that and just emit the signal with the new device name.

File: src/app/src/advancedsetupdialog.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    signals:
+        void sig_ni_dev_change();
+
+    private slots:
+        void on_confirm_new_advanced_setup();
+        void nidevice_indexChanged(int index);
+
+    private:
+        Ui::AdvancedSetupDialog *ui;
+        std::string new_niDev; //NI-DAQmx device name
```
Comment (jperson) [inline src/app/src/advancedsetupdialog.h:36]: I would call this one `m_niDev` and have it's value set in the constructor by reading the config file and then internally in this class set this to the user picked value and emit this variable in the signal

File: src/app/src/advancedsetupdialog.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        std::string* m_niDev;
```
Comment (jperson) [inline src/app/src/advancedsetupdialog.h:37]: I don't think you need this, definitely don't want to have a pointer back to the MainWindow class

File: src/app/src/advancedsetupdialog.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    Q_OBJECT
+
+    public:
+        explicit AdvancedSetupDialog(toml::value config,std::string* m_niDev, QWidget* parent = 0);
+        ~AdvancedSetupDialog();
+        void Initialize(std::vector<std::string>);
+        std::string Get_Device_Selected();
+
+    signals:
+        void sig_ni_dev_change();
```
Comment (jperson) [inline src/app/src/advancedsetupdialog.h:28]: this should take a parameter of a `std::string` and when emitting this signal pass in the `m_niDev` member variable

File: src/app/src/advancedsetupdialog.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        }
+        //save new ni device to toml file
+        std::filesystem::path configFile{fmt::format("{}/AppData/Local/Nautilus/nautilus.toml", userProfile.string())};
+        auto file = toml::parse(configFile);
+        file["device"]["nidaqmx"]["device"] = new_niDev;
+        std::ofstream outf;
+        outf.open(configFile.string());
+        outf << file << std::endl;
+        spdlog::info("Updated NI device name in toml");
+        emit this->sig_ni_dev_change();
```
Comment (jperson) [inline src/app/src/advancedsetupdialog.cpp:68]: I would update `m_niDev` here with the new value and then pass that to the signal

File: src/app/src/advancedsetupdialog.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+* When user selects a different ni device, save changes to be confirmed later.
+*
+* @param index of new choice
+*/
+void AdvancedSetupDialog::nidevice_indexChanged(int index){
+    new_niDev = ui->nidevicelist->currentText().toStdString();
+}
+
+
+std::string AdvancedSetupDialog::Get_Device_Selected(){
```
Comment (jperson) [inline src/app/src/advancedsetupdialog.cpp:86]: if you're using a signal I don't think this is necessary

File: src/app/src/mainwindow.cpp [hunk 9]
```diff
@@ -229,6 +230,29 @@ void MainWindow::Initialize() {
 }
 
 
+/*
+* Runs when a new ni device is selected, re configure ni device leds
+*/
+void MainWindow::Resetup_ni_device(){
```
Comment (jperson) [inline src/app/src/mainwindow.cpp:236]: this should take a `std::string` so the signal from advancedsetup can pass in the new device name

File: src/app/src/mainwindow.cpp [hunk 10]
```diff
@@ -339,9 +363,12 @@ void MainWindow::on_durationEdit_valueChanged(double value) {
 
 /*
  * Advanced setup button slot, called when user clicks on advanced setup button.
- * Currently does nothing.
  */
 void MainWindow::on_advancedSetupBtn_clicked() {
+    std::vector<std::string> devicelist = m_DAQmx.GetListOfDevices();
+    m_advancedSettingsDialog->Initialize(devicelist);
```
Comment (jperson) [inline src/app/src/mainwindow.cpp:369]: I think this and the `GetListOfDevices` should both be done in the constructor

**Review by jperson — APPROVED** [review]

**Pull request merged**